gh-106706: Streamline family syntax

kgdiem opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->
* Update bytecodes.c to use suggested syntax from #106706 
* Update checks in `generate_cases.py` to support >= 1 members and validate family name 
* Update documentation 
* Update tests 

<!-- gh-issue-number: gh-106706 -->
* Issue: gh-106706
<!-- /gh-issue-number -->

kgdiem committed:
gh-106706 Streamline family syntax

kgdiem committed:
Update syntax in interpreter_definition.md

kgdiem committed:
Further refactor bytecodes.c

kgdiem committed:
Update test

kgdiem committed:
Update syntax in interpreter_definition.md

kgdiem committed:
Update interpreter_definition to new syntax

kgdiem committed:
Use family name as first instruction

kgdiem committed:
stop using slice in iters

kgdiem committed:
Remove formatting in interpreter_definition.md

kgdiem committed:
Change test

kgdiem committed:
Merge branch 'main' into streamline-family-syntax

kgdiem commented on Tools/cases_generator/test_generator.py:
```diff
@@ -321,13 +321,13 @@ def test_macro_instruction():
                 _tmp_3 = res;
             }
             next_instr += 5;
-            static_assert(INLINE_CACHE_ENTRIES_OP == 5, "incorrect cache size");
```
It isn't clear to me how/where the INLINE_CACHE_ENTRIES_OP instruction is generated. 

Changing the test to pass w/o understanding doesn't inspire confidence on my end so any guidance or explanation here is greatly appreciated.

gvanrossum replied:
It is the second parameter to `family`.

kgdiem committed:
Fix check

kgdiem committed:
Fix check

kgdiem committed:
Fix assertion

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

kgdiem committed:
lint blurb

kgdiem committed:
Merge branch 'main' into streamline-family-syntax

markshannon committed:
GH-104909: Split `LOAD_ATTR_INSTANCE_VALUE` into micro-ops (GH-106678)

ammaraskar committed:
gh-106690: Add a .coveragerc file to the CPython repository (#8150)

The added file is the coverage default at some point in time + checking branches both ways + IDLE additions, labelled as such and somewhat designed to be unlikely to affect other files.  Located in the CPython repository directory, it can be used where it is or copied elsewhere, depending on how one runs coverage.
---------

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>
Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>
Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

gvanrossum committed:
gh-106701: Move the hand-written Tier 2 uops to bytecodes.c (#106702)

This moves EXIT_TRACE, SAVE_IP, JUMP_TO_TOP, and
_POP_JUMP_IF_{FALSE,TRUE} from ceval.c to bytecodes.c.

They are no less special than before, but this way
they are discoverable o the copy-and-patch tooling.

bdraco committed:
gh-106664: selectors: add get() method to _SelectorMapping (#106665)

It can be used to avoid raising and catching KeyError twice via __getitem__.

Co-authored-by: Inada Naoki <songofacandy@gmail.com>

nedbat committed:
docs: clarify Path.suffix (GH-106650)

erlend-aasland committed:
gh-106368: Increase Argument Clinic test coverage (#106728)

- improve output_parameter() coverage
- improve coverage for Function.kind

AlexWaygood committed:
gh-104683: Argument clinic: use an enum to describe the different kinds of functions (#106721)

Argument clinic: use an enum to describe the different kinds of functions

gvanrossum committed:
gh-106529: Split FOR_ITER_{LIST,TUPLE} into uops (#106696)

Also rename `_ITER_EXHAUSTED_XXX` to `_IS_ITER_EXHAUSTED_XXX` to make it clear this is a test.

gvanrossum commented on Misc/NEWS.d/next/Tools-Demos/2023-07-13-12-08-35.gh-issue-106706.29zp8E.rst:
```diff
@@ -0,0 +1,3 @@
+Change bytecode syntax for families
+to remove redundant name matching
+pseudo snytax
```
```suggestion
pseudo syntax.
```

gvanrossum commented on Python/executor_cases.c.h:
```diff
@@ -183,6 +182,7 @@
         }
 
         case TO_BOOL_ALWAYS_TRUE: {
+            static_assert(INLINE_CACHE_ENTRIES_TO_BOOL == 3, "incorrect cache size");
```
While it works, I think it's better to keep the static assert in the "family head".

gvanrossum replied:
This still isn't resolved?

kgdiem replied:
No, I think this might be related to the change in `Lib/_opcode_metadata.py`as well; the op from the family name isn't getting inserted into `instrs`

I'm having some trouble getting the name into `instrs` / parsed as an instruction. Any advice?

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -865,8 +865,13 @@ def check_families(self) -> None:
         """
         for family in self.families.values():
-            if len(family.members) < 2:
+            if len(family.members) < 1:
```
I don't think this can even happen any more -- if you try to create an empty family it will be a SyntaxError.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -877,10 +882,10 @@ def check_families(self) -> None:
                     f"Family {family.name!r} has unknown members: {unknown}", family
                 )
-            if len(members) < 2:
+            if len(members) < 1:
```
Ditto.

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -1303,11 +1308,11 @@ def write_metadata(self) -> None:
             self.out.emit("_specializations = {")
             for name, family in self.families.items():
-                assert len(family.members) > 1
+                assert len(family.members) >= 1
```
Ditto

gvanrossum commented on Tools/cases_generator/interpreter_definition.md:
```diff
@@ -351,7 +351,7 @@ A _family_ represents a specializable instruction and its specializations.
 
```
Perhaps the line above could say "A familiy *maps* a specializable instruction *to* ..." ?

gvanrossum reviewed (commented):
Thank you! I hope I've given you enough hints on what to do about the static assert -- if not, let me know.

sobolevn committed:
gh-105626: Change the default return value of `HTTPConnection.get_proxy_response_headers` (#105628)

grantramsay committed:
gh-105293: Do not call SSL_CTX_set_session_id_context on client side SSL context (#105295)

* gh-105293: Do not call SSL_CTX_set_session_id_context on client side SSL context

Openssl states this is a "server side only" operation.
Calling this on a client side socket can result in unexpected behavior

* Add news entry on SSL "set session id context" changes

CharlieZhao95 committed:
gh-106446: Fix failed doctest in stdtypes (#106447)

---------
Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

kgdiem committed:
Update Misc/NEWS.d/next/Tools-Demos/2023-07-13-12-08-35.gh-issue-106706.29zp8E.rst

Co-authored-by: Guido van Rossum <gvanrossum@gmail.com>

kgdiem committed:
Remove redundant checks for family member length

kgdiem committed:
Reword documentation

kgdiem committed:
Update test, regenerate headers

kgdiem committed:
keep formatting in interpreter_definiton

erlend-aasland committed:
Merge branch 'main' into streamline-family-syntax

gvanrossum commented on Lib/_opcode_metadata.py:
```diff
@@ -31,75 +31,3 @@
         "STORE_SUBSCR_DICT",
         "STORE_SUBSCR_LIST_INT",
     ],
-    "SEND": [
```
That this file changed surprises me. Maybe this points to a problem?

kgdiem replied:
It didn't change originally, I think only after https://github.com/python/cpython/pull/106716/commits/178e3009234e49102d2d8058504f9c25eb374a06 ?

gvanrossum commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -855,8 +855,11 @@ def check_families(self) -> None:
         for family in self.families.values():
-            if len(family.members) < 2:
-                self.error(f"Family {family.name!r} has insufficient members", family)
+            if family.name not in self.macro_instrs and family.name not in self.instrs:
+                self.error(
+                    f"Family {family.name!r} has unknown instruction {family.name!r}",
+                    family,
+                )
```
Good catch!

gvanrossum reviewed (commented):
I think you pushed a rebase? Please don't do that once review is in progress -- everything will be squashed when we merge into main.

Can you look into moving the static_assert back to the family head?

And I'm unsure why the metadata file was truncated. (Let me know if you need help figuring this out.)

kgdiem committed:
Regen cases

kgdiem committed:
Check against family name when emitting cache size check, add family.name to map

kgdiem committed:
update family syntax in test

kgdiem committed:
Regenerate cases

kgdiem committed:
Check families for mac.name instead of last_instr.family

gvanrossum reviewed (approved):
Perfect! I'll merge.
